### ISSUE TITLE
Random.NextDate

### ISSUE THREAD

**Issue by NUnitTester** [nunit/nunit#3817]
The randomizer is a great thing!

`TestContext.CurrentContext.Random`

The functionality to generate Random Dates in a defined range is missing.

`NextDate (DateTime startDate, DateTime endDate, Granularity granularity);` where granularity is a Enum that can be year, quarter, month, week, day, hour, minute, second

**Comment by rprouse** [comment]
Interesting idea. @nunit/framework-team any thoughts on this?

**Comment by Dreamescaper** [comment]
Sounds good, although I'm not sure about this Granularity enum. How it's going to work? What should happen if it's bigger than provided start-end interval? Do we need default value?

**Comment by stevenaw** [comment]
I like the idea of adding DateTime functionality to the Randomizer, and I could see some uses for `Granularity`, but I'd like to suggest the values there match the "units of time" fields exposed on the DateTime struct: Year, Month, Day, Hour, Minute, Second, Millisecond, Ticks. I'm hesitant about adding other granular units of time like Quarter. I'm concerned there could be edge cases or ambiguity around how they're interpreted, especially with regards to potential localization issues. Ex: Is quarter a fiscal quarter calendar quarter? In both cases, different locales may have different definitions of each.

I think it could be helpful with this feature to add a few overloads so that all the args could be optional (similar to .NET's `Random.Next()`).

I hadn't considered @Dreamescaper 's edge case around a large granularity and small delta between min/max. I'm not sure the best thing to do there

**Comment by rprouse** [comment]
I think `Granularity` is confusing. I think it means that a `Granularity` of `Seconds` means that the `DateTime` will have random values down to the second, but smaller will be zero?

I also agree with @stevenaw that the granularity should match the `DateTime` struct properties, `Quarter` is ambiguious. We should probably also have a default granularity of the smallest unit.

**Comment by stevenaw** [comment]
@nunit/framework-team 
The previous planning for this seemed to get "stuck" on the `Granularity`. I wonder, would `Precision` be a clearer term as a mathematical analog?

Should an exception be thrown if attempting @Dreamescaper 's edge case of a low precision (ex: year) with a narrow date range?

**Comment by manfred-brands** [comment]
@stevenaw The ticket doesn't define what `Granularity` means for the values.

Does `Granularity.Day` mean that all values are exactly days at 00:00:00.00
or that the values are one different days, but at random times during those days?

If the first, then `Granularity` is the right term.
If the second then `Interval` might be better.

I agree to use `DateTime` properties as the members.

For the edge cases:

`Random(DateTime.UtcNow, DateTime.UtcNow.AddDays(1), Granularity.Day)` means it can only return 1 Date (over and over again)
`Random(DateTime.UtcNow, DateTime.UtcNow.AddSeconds(1), Granularity.Day)` should throw an error.
```
if (end - start < TimeSpanFor[Granularity])
    throw ArgumentOutOfRangeException(...);
```

The above doesn't work for items like `Month` and `Year` as those could fluctuate, but should be close enough.

Instead of an enum for  `Granularity` we could use static fields:
```
public readonly struct Granularity
{
    public static readonly TimeSpan Day = new(TimeSpan.TicksPerDay);
    public static readonly TimeSpan Hour = new(TimeSpan.TicksPerHour);
    ...
};
```

**Comment by stevenaw** [comment]
Thanks @manfred-brands 
I think we'd interpreted `Granularity` as your first option: that the returned value is not more precise than requested. So passing `Minute` would return a value with the seconds zeroed out.

Though if you felt it unclear from the name "granularity" then perhaps that might indicate we should find a clearer term.

**Comment by stevenaw** [comment]
> means it can only return 1 Date (over and over again) should throw an error.

I agree this isn't useful. One counterpoint I could suggest for discussion: 

An API which returns a single value (at random) from a set of possible values should error for an empty set. A set with a single value may just return that value over and over. While it may not be "useful" it might still be considered "correct" from that perspective.

That would be behaviorally similar to how LINQ's `Single()` method works with the only differences that we are working with an unmaterialized set of possible values and that we are returning a random value instead of a deterministic one based on a predicate.

**Comment by manfred-brands** [comment]
That was what I was trying to say.
The first, despite not useful, is ok, the 2nd throws an error.
